### PR TITLE
Code Smells [Part I]

### PR DESCRIPTION
**Pull request by QilongTang**
Please Note:
1. Before submitting the PR, please review [How to Contribute to Dynamo](https://github.com/DynamoDS/Dynamo/blob/master/CONTRIBUTING.md)
2. Dynamo Team will meet 1x a month to review PRs found on Github (Issues will be handled separately)
3. PRs will be reviewed from oldest to newest
4. If a reviewed PR requires changes by the owner, the owner of the PR has 30 days to respond. If the PR has seen no activity by the next session, it will be either closed by the team or depending on its utility will be taken over by someone on the team
5. PRs should use either Dynamo's default PR template or [one of these other template options](https://github.com/DynamoDS/Dynamo/wiki/Choosing-a-Pull-Request-Template) in order to be considered for review.
6. PRs that do not have one of the Dynamo PR templates completely filled out with all declarations satisfied will not be reviewed by the Dynamo team.
7. PRs made to the `DynamoRevit` repo will need to be cherry-picked into all the DynamoRevit Release branches that Dynamo supports. Contributors will be responsible for cherry-picking their reviewed commits to the other branches after a `LGTM` label is added to the PR.

### Purpose

WIP for code smells in all submission within last two month. Want to give the disposal pattern more thoughts especially.

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [x] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@DynamoDS/dynamo 

### FYIs

(FILL ME IN, Optional) Names of anyone else you wish to be notified of

### CODE REVIEW

File: src/DocumentationBrowserViewExtension/DocumentationBrowserView.xaml.cs [hunk 1]
```diff
@@ -91,15 +91,22 @@ public void NavigateToPage(Uri link)
...
             this.documentationBrowser.DpiChanged -= DocumentationBrowser_DpiChanged;
         }
+
+        /// <summary>
+        /// Dispose function for DocumentationBrowser
+        /// </summary>
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
```
Comment (QilongTang) [inline src/DocumentationBrowserViewExtension/DocumentationBrowserView.xaml.cs:109]: @mjkkirschner Reminded me why we are not doing this. This make me wonder if it's worth to implement the Dispose pattern at all or just leave them as it is and look for a better solution.

File: src/DocumentationBrowserViewExtension/DocumentationBrowserViewModel.cs [hunk 2]
```diff
@@ -256,7 +262,7 @@ private string LoadContentFromResources(string name)
                 {
                     assemblyName = assemblyName.Substring(0, versionIndex);
                 }
-                assembly = AppDomain.CurrentDomain.GetAssemblies().Where(a => a.GetName().Name == assemblyName).FirstOrDefault();
+                assembly = AppDomain.CurrentDomain.GetAssemblies().FirstOrDefault(a => a.GetName().Name == assemblyName);
```
Comment (QilongTang) [inline src/DocumentationBrowserViewExtension/DocumentationBrowserViewModel.cs:265]: This is good point here from the report, in certain case I can see it saves time

File: src/DynamoCore/Configuration/DebugModes.cs [hunk 3]
```diff
@@ -134,7 +133,7 @@ public static DebugMode GetDebugMode(string name)
         public static bool IsEnabled(string name)
         {
             DebugMode dbgMode;
-            return debugModesEnabled && debugModes.TryGetValue(name, out dbgMode) ? dbgMode.IsEnabled : false;
+            return debugModesEnabled && debugModes.TryGetValue(name, out dbgMode) && dbgMode.IsEnabled;
```
Comment (QilongTang) [inline src/DynamoCore/Configuration/DebugModes.cs:136]: This is also something interesting I never give enough attention.

File: src/DocumentationBrowserViewExtension/DocumentationBrowserViewExtension.cs [hunk 4]
```diff
@@ -48,17 +48,18 @@ internal void OnMessageLogged(ILogMessage msg)
 
         #region IViewExtension lifecycle
 
-        public void Startup(ViewStartupParams viewLoadedParams)
+        public void Startup(ViewStartupParams p)
```
Comment (mmisol) [inline src/DocumentationBrowserViewExtension/DocumentationBrowserViewExtension.cs:51]: Doesn't it make more sense to put a better name in the interface? Single letter params are not a good practice.

File: src/DocumentationBrowserViewExtension/DocumentationBrowserViewModel.cs [hunk 5]
```diff
@@ -150,14 +156,14 @@ private void HandleLocalResource(OpenDocumentationLinkEventArgs e)
         {
             try
             {
-                var content = LoadContentFromResources(e.Link.ToString());
-                if (content == null)
+                var targetcontent = LoadContentFromResources(e.Link.ToString());
```
Comment (mmisol) [inline src/DocumentationBrowserViewExtension/DocumentationBrowserViewModel.cs:159]: NIT: Should be an upper case `C` like `targetContent`

File: src/DocumentationBrowserViewExtension/DocumentationBrowserViewExtension.cs (hunk 4 above)
Comment (QilongTang) [reply src/DocumentationBrowserViewExtension/DocumentationBrowserViewExtension.cs]: @mmisol I agree, will update

**Comment by QilongTang** [comment]
Running self CI

File: src/DocumentationBrowserViewExtension/DocumentationBrowserViewExtension.cs [hunk 6]
```diff
@@ -15,7 +15,7 @@ namespace Dynamo.DocumentationBrowser
     /// </summary>
     public class DocumentationBrowserViewExtension : IViewExtension, ILogSource
     {
-        private ViewLoadedParams viewLoadedParams;
+        private ViewLoadedParams viewLoadedParamsRefenrence;
```
Comment (mmisol) [inline src/DocumentationBrowserViewExtension/DocumentationBrowserViewExtension.cs:18]: Typo: There is an extra `n` in `Refenrence`

File: src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs [hunk 7]
```diff
@@ -192,25 +191,33 @@ public class HelixWatch3DViewModel : DefaultWatch3DViewModel
...
         private Dictionary<string, string> nodesSelected = new Dictionary<string, string>();
 
+        private Object element3DDictionaryMutex = new object();
+        /// <summary>
+        /// Mutex object
+        /// </summary>
+        public Object Element3DDictionaryMutex
+        {
+            get { return element3DDictionaryMutex; }
+        }
```
Comment (mmisol) [inline src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs:201]: From what I could see, the field was not being used outside of this class at all. Please, just keep the private field and delete the property. Mutex objects should not be shared publicly if possible.

Comment (QilongTang) [reply src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs:201]: That's true, I think it was a public property before that's why I added it. But API backward compatibility broken anyway when we renamed this :) 
@mjkkirschner How do you feel like we just remove this public property before, seems no one will be using it like @mmisol  noticed.

File: src/DocumentationBrowserViewExtension/DocumentationBrowserViewExtension.cs (hunk 6 above)
Comment (QilongTang) [reply src/DocumentationBrowserViewExtension/DocumentationBrowserViewExtension.cs:18]: Thanks, fixing

**Comment by QilongTang** [comment]
@mmisol Thanks for reviewing! I think the only thing left there are some TODOs in code saying making certain APIs private and a few spots saying we have too many params per function or function complexity too high. I will merge this for now and check the updated SQ report tomorrow morning just so we have a much cleaner baseline

**Comment by QilongTang** [comment]
All unit tests passing

**Pull request merged**